src/amqp-client/channel.cr

```diff
@@ -286,32 +304,51 @@ class AMQP::Client
       raise ClosedException.new(@closing_frame) if @closing_frame
     end
 
-    @on_confirm = Sync(Hash(UInt64, Proc(Bool, Nil))).new
+    @unconfirmed_publishes = Deque(UInt64).new
+    @unconfirmed_empty = ::Channel(Bool).new
+    @confirm_lock = Mutex.new
     @last_confirm = {0_u64, true}
 
+    @[Deprecated("Use `#wait_for_confirms` instead")]
     def on_confirm(msgid, &blk : Bool -> Nil)
       raise ArgumentError.new "Confirm id must be > 0" unless msgid > 0
       last_confirm, last_confirm_ok = @last_confirm
       if last_confirm >= msgid.to_u64
         blk.call last_confirm_ok
+      elsif last_unconfirmed = @unconfirmed_publishes.last
+        if last_unconfirmed >= msgid.to_u64
+          wait_for_confirms
+          blk.call last_confirm_ok
+        else
+          raise Error.new("msgid #{msgid} not published of this channel")
+        end
       else
-        @on_confirm[msgid.to_u64] = blk
+        wait_for_confirms
+        blk.call last_confirm_ok
       end
     end
 
     private def process_confirm(acked : Bool, delivery_tag : UInt64, multiple : Bool)
-      @last_confirm = {delivery_tag, acked}
-      if multiple
-        @on_confirm.reject! do |msgid, blk|
-          if msgid <= delivery_tag
-            blk.call acked
-            true
+      @confirm_lock.synchronize do
+        if @unconfirmed_publishes.first? == delivery_tag
+          @unconfirmed_publishes.shift
+        elsif idx = @unconfirmed_publishes.bsearch_index { |confirm_id| confirm_id >= delivery_tag }
+          if multiple
+            @unconfirmed_publishes.shift(idx + 1)
           else
-            false
+            @unconfirmed_publishes.delete_at(idx)
+          end
+        end
+        @last_confirm = {delivery_tag, acked}
+        if @unconfirmed_publishes.empty?
+          loop do
+            select
+            when @unconfirmed_empty.send acked
+            else
+              break
+            end
```

Comment:
how could `unconfirm_publishes` be empty here? that should be an error/exception, no?

Reply:
We just emptied it on the line just above?